### DISCUSSION TITLE
Azure Monitor (Logs): Allow to use template variables for resource groups and resource names

### DISCUSSION THREAD

**Discussion by andresmgot** [grafana/grafana#45133] [Data Sources]
Currently, it's not possible to use the template variables `ResourceNames` and `ResourceGroups` for Azure Monitor Logs. The resource picker only allows to select template variables for the whole path of the resource. Since the template variables listed above only return the name of the resource (or the group), these cannot be used.

The goal is to have the same template variable used for both Azure Monitor Metrics and Logs.

To achieve this we have several options:

Option 1. Allow to manually type the desired path in the resource picker. We could have an input in the resource picker for people to manually write the desired path. This path could include template variables:

![Screenshot from 2022-02-09 12-27-31](https://user-images.githubusercontent.com/4025665/153191950-511a7d37-ad3a-49f4-ac0d-e65930b64380.png)

Option 2a. Allow to select template variables for sub resources. We could display the same Template variable selector for all the nested resources. Allowing people to click a template variable under a resource group for example:

![Screenshot from 2022-02-09 12-40-25](https://user-images.githubusercontent.com/4025665/153193538-0280cede-9be3-491a-82ce-4cd11fe7e928.png)

Option 2b. Small variant from the previous one. We could filter out from the template variable selector variables that do not belong to the expanded resource (e.g. removing variables related to a resource group if we are selecting a resource name).

My preference is to go with 2a. Option 1 seems a bit verbose and with option 2b. we may end up still showing wrong variables (e.g. showing a template variable that points to a resource that is not present in the expanded resource group). Option 2a. is also the same approach that we have in Azure Monitor Metrics in which we allow people to select any template variable for any selector.

Thoughts?

**Comment by andresmgot** [comment]
related issue https://github.com/grafana/grafana/issues/43618

**Comment by sarahzinger** [comment]
hmm tough problem and a good write up!

With option 2, I would worry a little bit that it might be easy to select a template variable at the wrong level, for example you thought you were selecting a variable as a resource group but you picked it as a resource name because you missed that it was nested. For example in the screenshot you have, you could very easily misread that as a group instead of a name you know? We could fix the indentation to make it more obvious but I think people would still select the wrong level

I also have learned that people use template variables for partial names of resources (or at least this happens in metrics). So for example:
- people will have a template variable $environment
- then will want to write in a dropdown `some_shared_name_$environment`

I can imagine in option 1 how to do that but I can't imagine it in option 2

I'm open to whatever we decide, but I think my vote would be option 1, maybe hidden under some sort of advanced link that provides more options? Idk, seems worth chatting with UX if you haven't already they might have ideas we haven't thought of.

**Reply by andresmgot** [reply]
> you could very easily misread that as a group instead of a name you know? We could fix the indentation to make it more obvious but I think people would still select the wrong level

Right, the mock up can be a bit misplaced, in theory the indentation should be clear. Also, when you select something and close the resource picker, the resource should be shown as:

![Screenshot from 2022-02-09 14-42-29](https://user-images.githubusercontent.com/4025665/153213356-d7617687-2172-4745-9249-3f147e0d3724.png)

> then will want to write in a dropdown some_shared_name_$environment
> I can imagine in option 1 how to do that but I can't imagine it in option 2

Yeah, I still think that we may need to implement the option 1 but for "very" advance use cases. Using a template variable here should be somehow simple, that's why option 2 is a bit more user-friendly.

**Comment by andresmgot** [comment]
Singe Grafana 8.5, it's possible to use template variables and specify a resource URI using the resource Picker. In 9.0, the resource picker is also available for Metrics (previously only available for Logs).

<img width="985" alt="Screen Shot 2022-03-23 at 5 24 38 PM" src="https://user-images.githubusercontent.com/6620164/159798372-a1a6eb8a-77d2-4803-a879-d6f25c1334ff.png">

Please open a new issue if you still find problems with this!
